1-Draft/RFCNNNN-RFC-Native-Command-Exit-Errors.md

```diff
@@ -0,0 +1,236 @@
+---
+RFC: RFCNNNN-RFC-Native-Command-Exit-Errors
+Author: Micheal Padden
+Status: Draft
+Area: Engine, Language
+Version: 1.0.0
+---
+
+# Native Command Error Handling
+
+Although Powershell has an exception-based error handling framework, it
+does not currently apply to native commands. 
+
+Exception-based error handling makes it easier to write robust code
+as less boilerplate code is needed to check and handle errors.
+
+Powershell scripts using native commands would benefit from being able
+to error handling features like those used by cmdlets.
+
+## Motivation
+
+Native commands return an exit code to the calling application which
+will be zero for success or non-zero for failure. A robust script will
+not assume success, instead checking the exit code after any statement
+calling a native command, using boilerplate like the following:
+```Powershell
+ if( ! $? ) 
+ { 
+     exit $lastexitcode; 
+ }
+```
+
+Bourne type shells provide a `set -e` error handling mode that acts as 
+if this boilerplate were included in certain contexts. Like exception
+based error handling, this can exit a stack of functions, scripts and
+shells. This allows robust scripts to be written with a minimal amount
+of boilerplate.
+
+To support a similar style of error handling with native commands, this
+RFC proposes including native command errors in Powershell's exception
+handling framework in a similar way.
+
+The specification and alternative proposals are based on the
+[Equivalent of bash `set -e` #3415](https://github.com/PowerShell/PowerShell/issues/3415)
+and committee review of the associated
+[pull request](https://github.com/PowerShell/PowerShell/pull/3523).
+
+
+## Specification
+
+This RFC proposes including native commands in the error handling
+framework, by allowing an error to be reported to the error stream
+when a native command exits with a non-zero exit code, similar to 
+the `set -e` option in bourne type shells.
+
+The `$PSIncludeNativeCommandInErrorActionPreference` preference
+variable should govern treatment of non-zero exit codes on native
+commands. Possible values are:
+- `$false`: (the default) ignore non-zero exit codes. 
```

Comment:
I think rather than a boolean, a configuration like this is worth having enum values both for readability and extensibility

Reply:
Not convinced by this, because 
- The reason for the long name is to be unambiguous. 
- The reason for the boolean is for consistency with other flags. 
- My hope is that this discussion will settle on a firm solution for this issue, and avoid a stop-gap solution that needs to be extended again. If code is written to use this facility, I would want to behave appropriately without needing to revise that code to take advantage of a later fix to the fix....

Reply:
You mean other "flags" like `$ErrorActionPreference`?

Reply:
I suppose I should have said switch. I mean like the Log preference variables, which are switches. e.g. $LogCommandHealthEvent = False (not logged).

Reply:
If two named values are more powershelly for switches, that's cool. My concern really is making sure so far as possible that this will not have to be extended to fix some omission as happened with the unix shell version.

Reply:
This being software, it will probably need to be extended one day. Nothing is futurproof